lua/starfall/libs_sh/quaternion.lua

```diff
@@ -1,681 +1,837 @@
--- Global to all starfalls
 local checkluatype = SF.CheckLuaType
-
---[[
--- Quaternion Support
--- Converted from Wiremod's E2 Quaternion library for general lua use
--- Original code for use by Bubbus
--- Permission received for use from Bubbus by Radon
--- http:\\wiki.wiremod.com/?title=Expression2#Quaternion
---
--- Credits to Radon for addition to Starfall
--- Credits to Divran for painful amounts of testing
-]]
-
--- faster access to some math library functions
-local math = math -- Because global lookups suck
-local setmetatable = setmetatable
-local abs   = math.abs
-local Round = math.Round
-local sqrt  = math.sqrt
-local exp   = math.exp
-local log   = math.log
-local sin   = math.sin
-local cos   = math.cos
-local sinh  = math.sinh
-local cosh  = math.cosh
-local acos  = math.acos
-local min 	= math.min
-
-local delta = 0.0000001000000
-
-local deg2rad = math.pi / 180
-local rad2deg = 180 / math.pi
-
-
---- Quaternion library
--- @name quaternion
--- @class library
--- @libtbl quat_lib
-SF.RegisterLibrary("quaternion")
-
---- Quaternion type
+local dgetmeta = debug.getmetatable
+
+local math_sqrt = math.sqrt
+local math_exp = math.exp
+local math_log = math.log
+local math_sin = math.sin
+local math_cos = math.cos
+local math_min = math.min
+local math_acos = math.acos
+local math_clamp = math.Clamp
+local math_max = math.max
+local math_rad = math.rad
+local math_deg = math.deg
+
+
+-- Based on Expression's 2 quaternion library: https://github.com/wiremod/wire/blob/master/lua/entities/gmod_wire_expression2/core/quaternion.lua
+--- Quaternion type. Recently reworked, for full changelist visit: https://github.com/thegrb93/StarfallEx/pull/953
 -- @name Quaternion
 -- @class type
 -- @libtbl quat_methods
 -- @libtbl quat_meta
-SF.RegisterType("Quaternion")
-
-
-return function(instance)
-
-local quat_lib = instance.Libraries.quaternion
-local quat_methods, quat_meta = instance.Types.Quaternion.Methods, instance.Types.Quaternion
-local ent_meta, ewrap, eunwrap = instance.Types.Entity, instance.Types.Entity.Wrap, instance.Types.Entity.Unwrap
-local ang_meta, awrap, aunwrap = instance.Types.Angle, instance.Types.Angle.Wrap, instance.Types.Angle.Unwrap
-local vec_meta, vwrap, vunwrap = instance.Types.Vector, instance.Types.Vector.Wrap, instance.Types.Vector.Unwrap
-
-local getent
-instance:AddHook("initialize", function()
-	getent = instance.Types.Entity.GetEntity
+SF.RegisterType("Quaternion", true, false, nil, nil, function(checktype, quat_meta)
```

Comment:
This function arg is only needed for custom wrap/unwrap via glua types. You've already got your wrap/unwrap on line 44/48 so this can be deleted.

Reply:
I'll look at the rest later

Reply:
Done, I've also moved the helper functions out of the instance scope.